processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -0,0 +1,285 @@
+          fromKey.remaining(),
+          keySerde.keySize()
+      );
+    }
+
+    final int recordOffset = dimIndex * recordSize;
+
+    if (recordOffset + recordSize > buffer.capacity()) {
+      // This error cannot be recoverd, and the query must fail
+      throw new ISE(
+          "A record of size [%d] cannot be written to the array buffer at offset[%d] "
+          + "because it exceeds the buffer capacity[%d]. Try increasing druid.processing.buffer.sizeBytes",
+          recordSize,
+          recordOffset,
+          buffer.capacity()
+      );
+    }
+
+    if (!isUsedKey(buffer, recordOffset)) {
+      this.buffer.position(recordOffset);
+      this.buffer.putInt(Groupers.getUsedFlag(dimIndex));
+      this.buffer.put(fromKey);
+    }
+
+    final int baseOffset = recordOffset + keySize;
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].aggregate(buffer, baseOffset + aggregatorOffsets[i]);
+    }
+
+    return AggregateResult.ok();
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    // BufferArrayGrouper is used only for dictionary-indexed single-value string dimensions.
+    // Here, the key contains the dictionary-encoded value of the grouping key
+    // and we use it as the index for the aggregation array.
+    final ByteBuffer fromKey = keySerde.toByteBuffer(key);
+    final int dimIndex = fromKey.getInt() + 1; // the first index is for missing value
+    fromKey.rewind();
+    return aggregate(key, dimIndex);
+  }
+
+  private static boolean isUsedKey(ByteBuffer buffer, int pos)
+  {
+    return (buffer.get(pos) & 0x80) == 0x80;
+  }
+
+  @Override
+  public void reset()
+  {
+    buffer.putInt(0, 0); // for missing value
+    for (int i = 1; i < cardinality + 1; i++) {
+      buffer.putInt(i * recordSize, 0);
+      final int baseOffset = i * recordSize + keySize;
+      for (int j = 0; j < aggregators.length; ++j) {
+        aggregators[j].init(buffer, baseOffset + aggregatorOffsets[j]);
+      }
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  @Override
+  public Iterator<Entry<KeyType>> iterator(boolean sorted)
+  {
+    return sorted ? sortedIterator() : plainIterator();
+  }
+
+  private Iterator<Entry<KeyType>> sortedIterator()
+  {
+    // Sorted iterator is currently not used because there is no way to get grouping key's cardinality when merging
+    // partial aggregation result in brokers and even data nodes (historicals and realtimes).
+    // However, it should be used in the future.
+    final BufferComparator comparator = keySerde.bufferComparator();
+    final List<Integer> wrappedOffsets = IntStream.range(0, cardinality + 1).boxed().collect(Collectors.toList());
```

Comment:
I realized this is not needed because even the plain iterator returns entries in order of sorted keys.